lib/adm.js

```diff
@@ -4177,68 +4331,112 @@ maAdm.prototype.verifyPlan = function (opts, cb)
 		}
 	}
 
-	var imgapi_opts = {};
-	/*
-	 * The channel can be null when e.g. 'sdcadm channel unset' causes our
-	 * SAPI entry to have that metadata removed. In that case, when
-	 * searching for images, we look in the default channel as configured on
-	 * the updates imgapi server itself.
-	 */
-	if (self.ma_channel !== null) {
-		imgapi_opts.channel = self.ma_channel;
-	}
+	var fetch_images_opts = {
+	    log: self.ma_log,
+	    local_imgapi: self.ma_sdc.IMGAPI,
+	    remote_imgapi: self.ma_sdc.REMOTE_IMGAPI,
+	    services_by_image: services_by_image,
+	    channel: self.ma_channel,
+	    skip_verify_channel: opts.skip_verify_channel,
+	    skip_verify_remote: opts.skip_verify_remote
+	};
 
-	// An array of strings, later used to form an error message.
-	var imagesNotInTheChannel = [];
-	// process each of the images in parallel
-	vasync.forEachParallel({
-	    inputs: Object.keys(services_by_image),
-	    func: function channelCheckOneImage(imageUuid, nextImage) {
-		self.ma_log.debug(
-			    'looking up image %s for service %s on ' +
-			    'channel %s',
-			    imageUuid, services_by_image[imageUuid],
-			    self.ma_channel);
-		self.ma_sdc.REMOTE_IMGAPI.getImage(imageUuid, imgapi_opts,
-			function getImage(err, image) {
-			if (err) {
-				// note that we don't invoke the err callback
-				// since vasync will then report only
-				// 'the first of n errors'
-				var missing = sprintf('%s (%s)', imageUuid,
-				    services_by_image[imageUuid]);
-				self.ma_log.debug('missing image ' + missing);
-				imagesNotInTheChannel.push(missing);
-				nextImage();
+	// retrieve local and (optionally) remote Image objects in order to
+	// verify the plan.
+	fetchImages(fetch_images_opts,
+	    function verifyImages(err, local_images, remote_images) {
+		if (err) {
+			cb(err);
+			return;
+		}
+
+		// We don't have a Set() object, instead use unique Object keys
```

Comment:
Pro-tip: Even when there is a Set object, one thing to be aware of is that there is no natural JSON representation of a JS `Set`, which leads to a real pain in Bunyan JSON logging of them.  Obviously not relevant here right now.

Reply:
TIL - thanks!